File: app/src/main/java/com/phoneguard/blocker/BrowserWatcherService.java
package com.phoneguard.blocker;

import android.accessibilityservice.AccessibilityService;
import android.view.accessibility.AccessibilityEvent;
import android.view.accessibility.AccessibilityNodeInfo;
import java.util.HashSet;
import java.util.Set;

public class BrowserWatcherService extends AccessibilityService {
    private static final Set<String> browserPackages = new HashSet<String>();

    static {
        browserPackages.add("com.android.chrome");
        browserPackages.add("com.chrome.beta");
        browserPackages.add("org.mozilla.firefox");
        browserPackages.add("com.opera.browser");
        browserPackages.add("com.brave.browser");
        browserPackages.add("com.duckduckgo.mobile.android");
    }

    public void onAccessibilityEvent(AccessibilityEvent event) {
        String pkg = event.getPackageName() != null ? event.getPackageName().toString() : "";
        if (!browserPackages.contains(pkg)) return;

        if (GuardPrefs.isBlockActive(this) && !OverlayManager.isOverlayShowing()) {
            OverlayManager.showBlockOverlay(this);
            return;
        }
        if (OverlayManager.isOverlayShowing()) return;

        AccessibilityNodeInfo root = getRootInActiveWindow();
        if (root == null) return;

        Set<String> keywords = GuardPrefs.getKeywords(this);
        if (keywords.isEmpty()) return;

        String text = extractAllText(root).toLowerCase();

        for (String keyword : keywords) {
            if (text.contains(keyword.toLowerCase())) {
                OverlayManager.showBlockOverlay(this);
                return;
            }
        }
    }

    private String extractAllText(AccessibilityNodeInfo node) {
        if (node == null) return "";
        StringBuilder sb = new StringBuilder();
        Set<AccessibilityNodeInfo> visited = new HashSet<AccessibilityNodeInfo>();
        walk(node, sb, visited);
        return sb.toString();
    }

    private void walk(AccessibilityNodeInfo n, StringBuilder sb, Set<AccessibilityNodeInfo> visited) {
        if (n == null || visited.size() > 200) return;
        if (visited.contains(n)) return;
        visited.add(n);

        if (n.getText() != null && n.getText().length() > 0) {
            sb.append(n.getText()).append(" ");
        }
        if (n.getContentDescription() != null && n.getContentDescription().length() > 0) {
            sb.append(n.getContentDescription()).append(" ");
        }

        for (int i = 0; i < n.getChildCount(); i++) {
            walk(n.getChild(i), sb, visited);
        }
    }

    public void onInterrupt() {
    }
}

File: app/src/main/java/com/phoneguard/blocker/BrowserWatcherService.kt
package com.phoneguard.blocker

import android.accessibilityservice.AccessibilityService
import android.view.accessibility.AccessibilityEvent
import android.view.accessibility.AccessibilityNodeInfo

class BrowserWatcherService : AccessibilityService() {

    private val browserPackages = setOf(
        "com.android.chrome",
        "com.chrome.beta",
        "org.mozilla.firefox",
        "com.opera.browser",
        "com.brave.browser",
        "com.microsoft.emmx",
        "com.sec.android.app.sbrowser",
        "com.duckduckgo.mobile.android"
    )

    override fun onAccessibilityEvent(event: AccessibilityEvent) {
        val pkg = event.packageName?.toString() ?: return
        if (pkg !in browserPackages) return

        if (GuardPrefs.isBlockActive(this) && !OverlayManager.isOverlayShowing()) {
            OverlayManager.showBlockOverlay(this)
            return
        }
        if (OverlayManager.isOverlayShowing()) return

        val root = rootInActiveWindow ?: return
        val keywords = GuardPrefs.getKeywords(this)
        if (keywords.isEmpty()) return

        val text = extractAllText(root)
        val lowerText = text.lowercase()

        val matched = keywords.any { lowerText.contains(it) }
        if (matched) {
            OverlayManager.showBlockOverlay(this)
        }
    }

    private fun extractAllText(node: AccessibilityNodeInfo?): String {
        if (node == null) return ""
        
        val sb = StringBuilder()
        val visited = mutableSetOf<AccessibilityNodeInfo>()
        
        fun walk(n: AccessibilityNodeInfo?) {
            if (n == null || visited.size > 200) return
            if (visited.contains(n)) return
            visited.add(n)
            
            n.text?.let { if (it.isNotEmpty()) sb.append(it).append(" ") }
            n.contentDescription?.let { if (it.isNotEmpty()) sb.append(it).append(" ") }
            
            for (i in 0 until n.childCount) {
                walk(n.getChild(i))
            }
        }
        
        walk(node)
        return sb.toString()
    }

    override fun onInterrupt() {}
}
